### PR TITLE
Update TypeMismatch for multiple possible types.

### PR DESCRIPTION
**Pull request by bsermons**
Per comment in https://github.com/paf31/purescript-foreign-generic/pull/11

### CODE REVIEW

File: src/Data/Foreign.purs [hunk 1]
```diff
@@ -44,16 +44,20 @@ foreign import data Foreign :: *
...
   | JSONError String
 
 instance showForeignError :: Show ForeignError where
-  show (TypeMismatch exp act) = "Type mismatch: expected " <> exp <> ", found " <> act
   show (ErrorAtIndex i e) = "Error at array index " <> show i <> ": " <> show e
   show (ErrorAtProperty prop e) = "Error at property " <> show prop <> ": " <> show e
   show (JSONError s) = "JSON error: " <> s
+  show (TypeMismatch exps act) = "Type mismatch: expected " <> to_s exps <> ", found " <> act
+    where
+      to_s [] = "???"
```
Comment (paf31) [inline src/Data/Foreign.purs]: I think we need to use a non-empty array, for this reason.

Comment (garyb) [inline src/Data/Foreign.purs]: I agree, but just to say, `purescript-nonempty` is in contrib currently and this is core.

File: src/Data/Foreign.purs [hunk 2]
```diff
@@ -44,16 +44,20 @@ foreign import data Foreign :: *
 
 -- | A type for runtime type errors
 data ForeignError
-  = TypeMismatch String String
+  = TypeMismatch (Array String) String
   | ErrorAtIndex Int ForeignError
   | ErrorAtProperty String ForeignError
   | JSONError String
 
 instance showForeignError :: Show ForeignError where
```
Comment (paf31) [inline src/Data/Foreign.purs]: While we're breaking things here, we should make this into a valid `Show` instance, and move this into a `renderForeignError` function.

**Comment by paf31** [comment]
Thanks!

**Comment by paf31** [comment]
An alternative here is to move the non-empty array to the outer level:

```haskell
type F = Either (NonEmpty ForeignError)
``` 

We probably don't want a dependency on any non core library for `NonEmpty` though, so we should think carefully about that.

File: src/Data/Foreign.purs (hunk 2 above)
Comment (bsermons) [inline src/Data/Foreign.purs]: I'm not sure I understand, what makes this an invalid `Show` instance?

Comment (garyb) [inline src/Data/Foreign.purs]: There's an informal rule that `Show` should print purescript code, so like `"(TypeMismatch " <> show exp <> show act <> ")"` style, for each case.

**Comment by bsermons** [comment]
I can see the value of adding `NonEmpty` at the top level since it would make it good for use with validating the whole structure but is there a reason it couldn't exist at the `TypeMismatch` level also? 
Seems like it would be hard to get a nice message like the 'expected one of [...], found ...'  with it only at the top level.

What are the requirements for a package to be in core vs contrib?

File: src/Data/Foreign.purs (hunk 2 above)
Comment (bsermons) [inline src/Data/Foreign.purs]: Ah I see, thanks.

**Comment by paf31** [comment]
The `Show` instance looks good thanks. I think we just need to come to some conclusion regarding `NonEmpty`.

**Comment by bsermons** [comment]
Yea I just did the easy thing first since I wasn't sure if `NonEmpty` being in contrib is a show stopper or if there was some other alternative than to using `NonEmpty`.  

I'll continue ahead with it if you don't think it'll be a problem and see how far I can get with the suggestion above with moving the dependency to the outer level.

**Comment by bsermons** [comment]
Played around with using `type F = Either (NonEmpty ForeignError)`. You can see what it would look like [here (purescript-foreign)](https://github.com/bsermons/purescript-foreign/tree/nonempty-errors) and [here (purescript-foreign-generic)](https://github.com/bsermons/purescript-foreign-generic/tree/nonempty-errors).  

The last commit on each is with `TypeMismatch String String` and the parent commit is with `TypeMismatch (Array String) String`.

Some notes:
End up using `unsafePartial` to go from `SigProd`'s constructors array to `NonEmpty`
Wasn't sure what to use as the expected type when `SigProd`'s constructors were empty, so just used the name.
Lose the nicer message when using the simple `String String` constructor.

**Pull request closed without merging**